### PR TITLE
[RFC] feat(context): allow tags to have an optional value

### PR DESCRIPTION
**Pull request by raymondfeng**
This PR enhances our binding tags to allow optional values. For example, we can associate a `controller` binding with tags such as:
- 'controller' (tag name only)
- {name: 'MyController'} (tag with name and value)
- {extensionPoint: 'controllers'}

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/context/src/binding.ts [hunk 1]
```diff
@@ -215,17 +215,47 @@ export class Binding<T = BoundValue> {
...
-      });
+  /**
+   * Tag the binding with names
+   * @param tagNames A list of names
+   *
+   * @example
+   * ```ts
+   * binding.tag('a');
+   * binding.tag('a', 'b');
+   * binding.tag(['a', 'b'], 'c')
```
Comment (bajtos) [inline packages/context/src/binding.ts]: I find this syntax confusing. What does `['a', 'b']` mean? Is it creating a tag with name `a` and value `b`? Can we use an object with named properties instead please?

```ts
// add a single tag with no value (backwards compatiblity)
binding.tag('controller');
// add any number of tags with no value - do we want this?
binding.tag('controller', 'rest');
// add tags with values
binding.tag({controller: true, 'extensionPoint': 'controllers'});
```

Maybe your intention was different - could you please provide more details?

File: packages/context/src/binding.ts [hunk 2]
```diff
@@ -215,17 +215,47 @@ export class Binding<T = BoundValue> {
...
+  tagValue(name: string, value: BoundValue): this {
+    this.tagValues.set(name, value);
+    return this;
+  }
+
+  /**
+   * Get a set of tag names
+   */
+  get tags() {
+    return new Set<string>(this.tagValues.keys());
```
Comment (bajtos) [inline packages/context/src/binding.ts]: Is there any particular reason why you are converting the result of `tagValues.keys()` to a `Set`? Cannot we simply return whatever `keys` returned and let the caller decide how they want to work with that? Allocating a new Set whenever we are querying binding tags may hurt performance.

File: packages/context/test/unit/binding.unit.ts [hunk 3]
```diff
@@ -52,6 +52,28 @@ describe('Binding', () => {
       expect(binding.tags.has('t1')).to.be.true();
       expect(binding.tags.has('t2')).to.be.true();
     });
+
+    it('tags the binding with rest args', () => {
+      binding.tag('t1', ['t2', 't3'], 't4');
+      expect(binding.tags.has('t1')).to.be.true();
+      expect(binding.tags.has('t2')).to.be.true();
+      expect(binding.tags.has('t3')).to.be.true();
+      expect(binding.tags.has('t4')).to.be.true();
```
Comment (bajtos) [inline packages/context/test/unit/binding.unit.ts]: When one of the assertion fails, the error message is rather unhelpful: _expected false to be true_.

Please rewrite the test to assert on the array of tag keys, something along the following lines:

```ts
expect(binding.tags.keys()).toContainUnordered(['t1', 't2', 't3', 't4']);
```

File: packages/context/test/unit/binding.unit.ts [hunk 4]
```diff
@@ -52,6 +52,28 @@ describe('Binding', () => {
...
+    });
+  });
+
+  describe('tagValue', () => {
+    it('tags the binding with value', () => {
+      binding.tagValue('t1', 'val1');
+      expect(binding.tags.has('t1')).to.be.true();
```
Comment (bajtos) [inline packages/context/test/unit/binding.unit.ts]: Similarly here. Can you convert `tags` map to a regular object and then use `deepEqual` assertion?

```ts
expect(/*convert tags to an object*/).to.deepEqual({t1: 'val1'});
```

File: packages/context/test/unit/binding.unit.ts [hunk 5]
```diff
@@ -52,6 +52,28 @@ describe('Binding', () => {
...
+      expect(binding.tagValues.get('t1')).to.equal('val1');
+    });
+
+    it('tags the binding with a name', () => {
+      binding.tag('t1');
+      expect(binding.tags.has('t1')).to.be.true();
+      expect(binding.tagValues.get('t1')).to.equal(true);
```
Comment (bajtos) [inline packages/context/test/unit/binding.unit.ts]: Ditto.

```ts
expect(/*convert tags to an object*/).to.deepEqual({t1: true});
```

Please improve the test name to mention what value is being used, e.g.

```ts
it('sets the tag value to true when tagging by a name only');
```

File: packages/context/test/unit/context.unit.ts [hunk 6]
```diff
@@ -272,6 +272,24 @@ describe('Context', () => {
     });
   });
 
+  describe('findByTagValue', () => {
+    it('returns matching binding', () => {
+      const b1 = ctx.bind('foo').tagValue('t1', 'val1');
+      ctx.bind('bar').tag('t2');
+      ctx.bind('baz').tagValue('t3', 1);
```
Comment (bajtos) [inline packages/context/test/unit/context.unit.ts]: Would you mind using a more descriptive strings than bar/baz/t1/t2?

For example:
 - foo => ProductController
 - bar => ProductRepository
 - baz => db
 - t1 => controller
 - t2 => repository
 - t3 => datasource

File: packages/context/src/binding.ts (hunk 1 above)
Comment (raymondfeng) [reply packages/context/src/binding.ts]: The first round was to keep `tag` method only for names using rest params. It allows string or string[]. A new `tagValue` method was introduced to set name/value for tags.

With your feedback, I have consolidated both flavors into `tag` and allow tags to be set by name/value objects.

File: packages/context/src/binding.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/context/src/binding.ts]: It was `Set` before to avoid duplicate entries. Now we can use `string[]` as the type.

File: packages/context/test/unit/binding.unit.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/context/test/unit/binding.unit.ts]: Fixed.

File: packages/context/test/unit/binding.unit.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/context/test/unit/binding.unit.ts]: Fixed.

File: packages/context/src/binding.ts [hunk 7]
```diff
@@ -99,7 +99,7 @@ export enum BindingType {
 
 export class Binding<T = BoundValue> {
   public readonly key: string;
-  public readonly tags: Set<string> = new Set();
+  public readonly tagValues: {[name: string]: BoundValue} = {};
```
[the remark this answers is not in the document]
Comment (bajtos) [reply packages/context/src/binding.ts]: I also find `tagValues` a bit confusing - I'd expect `b.tagValues` to return only the values associated with tags without the tag names, similarly to what `Object.values()` does.

Ideally, our API should follow (be similar to) the ES2017 API for accessing keys, values and entries:
 - `Object.keys()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys
 - `Object.values()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/values
 - `Object.entries()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

Perhaps we should rename this property to something like `tagMap`?

File: packages/context/test/unit/context.unit.ts [hunk 8]
```diff
@@ -272,6 +272,24 @@ describe('Context', () => {
     });
   });
 
+  describe('findByTagValues', () => {
+    it('returns matching binding', () => {
+      const b1 = ctx.bind('foo').tag({name: 'my-controller'});
+      ctx.bind('bar').tag('t2');
+      ctx.bind('baz').tag({weight: 1});
```
Comment (bajtos) [inline packages/context/test/unit/context.unit.ts]: Reposting https://github.com/strongloop/loopback-next/pull/1262#discussion_r182082629:

Would you mind using a more descriptive strings than bar/baz/t1/t2?

For example:
 - foo => ProductController
 - bar => ProductRepository
 - baz => db
 - t2 => repository
 - t3 => datasource

**Comment by raymondfeng** [comment]
@bajtos Comments addressed. PTAL.

File: packages/context/src/context.ts [hunk 9]
```diff
@@ -185,19 +190,37 @@ export class Context {
...
       typeof pattern === 'string' ? this.wildcardToRegExp(pattern) : pattern;
-    return this.find(b => Array.from(b.tags).some(t => regexp.test(t)));
+    return this.find(b => Array.from(b.tagNames).some(t => regexp.test(t)));
+  }
+
+  /**
+   * Find bindings by tag name/value pairs
+   * @param tagNameValueMap Tag name/value pairs
+   */
+  findByTagValues<ValueType = BoundValue>(
```
[the remark this answers is not in the document]
Comment (virkt25) [reply packages/context/src/context.ts]: I think if we can, we should only have one method for finding a binding by tags ... `findByTag`

File: packages/context/test/acceptance/tagged-bindings.feature.md [hunk 10]
```diff
@@ -20,7 +20,48 @@ let ctx = new Context();
...
+
+```ts
+// create a container for bindings
+let ctx = new Context();
+
+// create a tagged binding
+let binding = ctx
+  .bind('foo')
+  .to('bar')
+  .tag({name: 'my-controller'});
```
Comment (shimks) [inline packages/context/test/acceptance/tagged-bindings.feature.md]: I think this should be `.tag({name: 'my-controller'}, 'controller')`

File: packages/context/src/binding.ts [hunk 11]
```diff
@@ -215,17 +253,53 @@ export class Binding<T = BoundValue> {
...
+   * // Add two tags
+   * // - `controller` (name = 'controller')
+   * // `{name: 'my-controller'}` (name = 'name', value = 'my-controller')
+   * binding.tag('controller', {name: 'my-controller'});
+   *
+   * // Add two named tags: `controller` and `rest`
+   * binding.tag(['controller', 'rest']);
+   * ```
+   */
+  tag(...tagNamesOrValues: (string | string[] | object)[]): this {
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/context/src/binding.ts]: I want to allow the mixed use of flavors, for example:
```
binding.tag('controller', 'rest');
binding.tag('controller', {name: 'my-controller'});
```
For string[], `MapObject<any>` does not exclude `Array` and I cannot find a way to disallow array at compile time. That's why I made api allow `string[]` as an arg:
```
binding.tag(['controller', 'rest']);
```

File: packages/context/test/acceptance/tagged-bindings.feature.md (hunk 10 above)
Comment (raymondfeng) [reply packages/context/test/acceptance/tagged-bindings.feature.md]: Fixed

**Comment by raymondfeng** [comment]
All comments are addressed. PTAL.

**Comment by jannyHou** [comment]
@virkt25 I guess the reason behind supporting a `tagMap` is to add tags in a more organized way, for example we add two tags for a binding, one specifies the location and the other one specifies the provider, without `tagMap` the biding would be 
`ctx.bind('info').to({...someInfo}).tag('New York', 'Whole Foods')` 
`ctx.bind('info').to({...someOtherInfo}).tag('Toronto', 'Walmart')`

This PR organize the tags in a way that people who add tags and who find the binding value by filtering them will have a better understanding what those tags mean: 

`ctx.bind('info').to({...someInfo}).tag({location: 'New York'}, {provider: 'Whole Foods'})` 
`ctx.bind('info').to({...someOtherInfo}).tag({location: 'Toronto'}, {provider: 'Walmart'})`

**Comment by virkt25** [comment]
Thanks for the explanations! I'm not opposed to the idea of a map, just thought we can do the same by just searching for a tag like `authentication.strategies` but organization definitely helps.

File: packages/context/src/binding.ts [hunk 12]
```diff
@@ -90,17 +91,55 @@ export enum BindingScope {
...
+   */
   CLASS = 'Class',
+  /**
+   * A provider class with `value()` function to get the value
+   */
   PROVIDER = 'Provider',
 }
 
+// tslint:disable-next-line:no-any
+export type TagMap = MapObject<any>;
```
Comment (virkt25) [inline packages/context/src/binding.ts]: Should `MapObject`'s generic type be of type `string` and not `any`?

Comment (raymondfeng) [reply packages/context/src/binding.ts]: No. `any` is for tag values, which can be a string, number, boolean, or even objects. The keys of TagMap are tag names (string).

Comment (virkt25) [reply packages/context/src/binding.ts]: Oh, ok. I didn't realize we supported tag values of types other than `string`.

**Comment by bajtos** [comment]
Let's continue the discussion from a collapsed comment https://github.com/strongloop/loopback-next/pull/1262#discussion_r182726005 here:

> > I would personally prefer a simpler API:
> > ```ts
> > tag(...tagNames: string[]);
> > tag(tagNamesWithValues: MapObject<any>);
> > ```

> I want to allow the mixed use of flavors, for example:
> ```ts
> binding.tag('controller', 'rest');
> binding.tag('controller', {name: 'my-controller'});
> ```

With my proposal in place, this can be easily achieved by calling `.tag` twice:

```ts
binding.tag('controller', 'rest').tag({name: 'my-controller'});
```

I see my proposed API as simpler to understand by people reading the code calling `.tag()`, and also simpler to implement & maintain.

> For `string[]`, `MapObject<any>` does not exclude `Array` and I cannot find a way to disallow array at compile time. That's why I made api allow `string[]` as an arg:
> ```ts
> binding.tag(['controller', 'rest']);
> ```

I am afraid I don't understand what is the use case and the problem here :(

Isn't the effect of calling `binding.tag(['controller', 'rest'])` the same as `binding.tag('controller', 'rest')`? If you are worried about the use case where the array of tag names is stored in a variable/function argument, then this can be easily solved by a spread operator:

```ts
const tagNames = ['controller', 'rest'];
binding.tag(...tagNames);
```

Or were you perhaps explaining why you are using `object` instead of `MapObject<any>`? Doesn't `object` allow arrays of arbitrary item type, e.g. `[1, 2, new Date()]` too?

**Comment by virkt25** [comment]
> With my proposal in place, this can be easily achieved by calling .tag twice:

Can we not further simplify it by allowing the following:
`controller.tag('rest', 'controller', {name: 'UserController'})`

**Comment by raymondfeng** [comment]
@bajtos 

1. Since we allow tags to be a simple `string` or key/value pair, I prefer to allow `binding.tag` to accept either form. For example:
```ts
binding.tag('controller', {name: 'my-controller'}); // It's a short form of binding.tag('controller').tag({name: my-controller});
```

2. I was referring to the fact that `MapObject<any>` allows array values. As a result, the following usage is valid even with your proposed signature:
```ts
binding.tag(['controller', 'rest']);
```
We have two options here:
  - Check the value and throw an error if it's an array.
   - Allow array values as I have implemented in the PR.

In summary, it boils down to two questions:

1. Do we allow `binding.tag('controller', {name: 'my-controller'})`? // Mixed forms of tags in one call
2. Do we allow `binding.tag(['controller', 'rest'])`? // An array of tags as a parameter

**Comment by bajtos** [comment]
@raymondfeng thank you for the explanation!

> Do we allow `binding.tag('controller', {name: 'my-controller'})`? // Mixed forms of tags in one call

I personally think this form is not worth the extra complexity it adds. However, if the rest of the team thinks it's a useful sugar API then I am ok with allowing it.

> Do we allow `binding.tag(['controller', 'rest'])`? // An array of tags as a parameter

I find this very confusing, especially when the array has exactly two items. It looks like a key-value pair to me, therefore I am never sure whether such code is setting two tags (`controller` and `rest`), or whether it's setting a single tag `controller` to a value `rest`.

> I was referring to the fact that MapObject<any> allows array values. As a result, the following usage is valid even with your proposed signature.

Uh, I see what you mean now! I find this surprising, although I understand why it works that way. The joys of JavaScript :)

**Comment by bajtos** [comment]
> We have two options here:
>
> - Check the value and throw an error if it's an array.
> - Allow array values as I have implemented in the PR.

I personally prefer the first option - check the value and throw an error if it's an array.

**Comment by raymondfeng** [comment]
> I personally prefer the first option - check the value and throw an error if it's an array.

@bajtos I made the change to throw an error if one of the arguments is an array.

PTAL.
